<launch>
    <arg name="quiet" default="false"/>
    <arg name="real_world" default="false"/>

<!-- Exploration -->
    <node pkg="exploration" type="filter_scan.py" name="filter_scan_node" output="screen">
        <param name="use_sim_time" value="false" />
        <param name="quiet" value="$(arg quiet)"/>
    </node>

    <node pkg="exploration" type="astar_planner_py.py" name="astar_planner_node" output="screen">
        <param name="use_sim_time" value="false" />
        <param name="map_topic" value="/mapUGV" />
        <param name="occupancy_thresh" value="50" />
        <param name="inflate_radius" value="0.0" />
        <param name="quiet" value="$(arg quiet)"/>
    </node>


<!-- Real World -->
    <group if="$(arg real_world)">
        <arg name="navigation_goal_topic" default="/move_base_simple/goal"/>
        <remap from="/camera/image/compressed" to="/theta_driver_node/image/compressed"/>
        <remap from="/state_estimation" to="/Odometry"/>
        <remap from="/state_estimation_at_scan" to="/Odometry"/>
        <remap from="/way_point_with_heading" to="/way_point_with_heading"/>

    <!-- Map Conversion -->
        <node pkg="exploration" type="occupancy_map.py" name="occupancy_map_node" output="screen">
            <param name="quiet" value="$(arg quiet)"/>
        </node>

    <!-- Exploration -->
        <node pkg="exploration" type="360_clip.py" name="clip_node" output="screen">
            <param name="use_sim_time" value="false" />
            <param name="quiet" value="$(arg quiet)"/>
        </node>
        <node pkg="exploration" type="frontier_mapping.py" name="frontier_mapping_node" output="screen">
            <param name="use_sim_time" value="false" />
            <param name="quiet" value="$(arg quiet)"/>
            <param name="real_world" value="$(arg real_world)"/>
        </node>
        <node pkg="exploration" type="object_navigation.py" name="object_navigation_node" output="screen">
            <param name="use_sim_time" value="false" />
            <param name="quiet" value="$(arg quiet)"/>
            <param name="real_world" value="$(arg real_world)"/>
        </node>

        <node pkg="exploration" type="path_follower.py" name="path_follower" output="screen">
            <param name="quiet" value="$(arg quiet)"/>
        </node>

        <node pkg="exploration" type="multi_view.py" name="multi_view" output="screen">
            <param name="quiet" value="$(arg quiet)"/>
            <param name="local_grid_resolution" value="0.5"/>
            <param name="real_world" value="$(arg real_world)"/>
        </node>

        <!-- Coverage Path Planning -->
        <node pkg="exploration" type="coverage_path.py" name="coverage_path_planner" output="screen">
            <param name="quiet" value="$(arg quiet)"/>
            <param name="grid_n" value="3"/>
            <param name="grid_m" value="3"/>
            <param name="local_grid_resolution" value="0.05"/>
            <param name="grid_margin" value="0.2"/>
            <param name="origin_filter_distance" value="3.0"/>
        </node>

    </group>

<!-- Simulation -->
    <group unless="$(arg real_world)">
    <!-- Map Conversion -->
        <!-- 3D→2D 맵 변환: 지면 & 슬로프 파라미터 -->
        <!-- <node pkg="mapconversion" type="map_conversion_oct_node" name="map_conversion" output="screen">
            <remap from="octomap" to="octomap_binary"/> -->
            <!-- 바닥 노이즈 제거: 10cm 이하 무시 -->
            <!-- <param name="minimum_z"             value="0.2"/> -->
            <!-- 최소 관측 횟수: 3~5회 -->
            <!-- <param name="minimum_occupancy"     value="50"/>
            <param name="map_frame"             value="map"/> -->
            <!-- UGV 주행 불가 경사: 0.3rad (≈17°) 이상 -->
            <!-- <param name="max_slope_ugv"         value="0.1"/> -->
            <!-- 경사 산출 윈도우 크기: 3 -->
            <!-- <param name="slope_estimation_size" value="1.5"/>
        </node> -->

        <!-- OctoMap 서버: 해상도 및 센서 최대거리 -->
        <!-- <node pkg="octomap_server" type="octomap_server_node" name="octomap_server" output="screen">
            <param name="resolution"          value="0.1"/>
            <param name="sensor_model/max_range" value="5.0"/>
            <param name="frame_id"            type="string" value="map"/>
            <param name="base_frame_id"       type="string" value="baselink"/>
            <param name="occupancy/meanUpdate" value="false"/>
            <param name="publish_free_space" value="true"/>
            <param name="publish_occupied_nodes" value="true"/>
            <param name="minimum_occupancy" value="80"/> -->
            <!-- 로봇 높이를 고려해, 너무 낮은 z 축 지점 무시 -->
            <!-- <param name="minimum_z"         value="0.05"/>
            <param name="maximum_z"         value="1.2"/> -->
            <!-- 경사 허용 범위 조정(지면 외부 역노이즈 억제) -->
            <!-- <param name="max_slope_ugv"     value="0.3"/>
            <remap from="cloud_in" to="/sensor_scan"/>
        </node> -->

    <!-- Map Conversion -->
        <node pkg="exploration" type="occupancy_map.py" name="occupancy_map_node" output="screen">
            <param name="quiet" value="$(arg quiet)"/>
        </node>

    <!-- Exploration -->
        <node pkg="exploration" type="360_clip.py" name="clip_node" output="screen">
            <param name="use_sim_time" value="false" />
            <param name="quiet" value="$(arg quiet)"/>
        </node>
        <node pkg="exploration" type="frontier_mapping.py" name="frontier_mapping_node" output="screen">
            <param name="use_sim_time" value="false" />
            <param name="quiet" value="$(arg quiet)"/>
            <param name="real_world" value="$(arg real_world)"/>
        </node>
        <node pkg="exploration" type="path_follower.py" name="path_follower" output="screen">
            <param name="quiet" value="$(arg quiet)"/>
        </node>
        <node pkg="exploration" type="object_navigation.py" name="object_navigation_node" output="screen">
            <param name="use_sim_time" value="false" />
            <param name="quiet" value="$(arg quiet)"/>
            <param name="real_world" value="$(arg real_world)"/>
        </node>

        <!-- Coverage Path Planning -->
        <node pkg="exploration" type="coverage_path.py" name="coverage_path_planner" output="screen">
            <param name="quiet" value="$(arg quiet)"/>
            <param name="grid_n" value="3"/>
            <param name="grid_m" value="3"/>
            <param name="local_grid_resolution" value="0.05"/>
            <param name="grid_margin" value="0.2"/>
            <param name="origin_filter_distance" value="3.0"/>
        </node>

    </group>

</launch>
